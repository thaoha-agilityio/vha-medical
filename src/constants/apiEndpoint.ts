export const API_ENDPOINT = {
  AUTH: '/auth/local',
  APPOINTMENTS: '/appointments',
  NOTIFICATIONS: '/notifications',
  SPECIALTIES: '/specialties',
  USERS: '/users',
  UPLOAD: '/upload',
  CHEMISTS: '/chemists',
  PERMISSIONS: '/users-permissions',
};

export const API_URL = `${process.env.NEXT_PUBLIC_API_URL}/api`;
export const IMGBB_URL = `${process.env.NEXT_PUBLIC_IMGBB_URL}?key=${process.env.NEXT_PUBLIC_IMGBB_API_KEY}`;

export const AUTH_TOKEN = process.env.NEXT_PUBLIC_AUTH_TOKEN;
