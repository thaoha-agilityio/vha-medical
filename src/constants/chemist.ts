import { UserModel } from '@/types';

export const DEFAULT_CHEMIST_DATA: UserModel = {
  email: '',
  username: '',
  password: '',
  description: '',
  rating: 0,
  tasks: 0,
  reviews: 0,
  specialtyId: undefined,
  avatar: undefined,
};

export const CHEMISTS_SEARCH_PARAMS = ['specialtyId'] as const;
