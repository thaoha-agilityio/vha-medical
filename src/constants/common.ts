export const RESULT_NOT_FOUND = 'Result Not Found';
