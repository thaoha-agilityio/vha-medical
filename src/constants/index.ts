export * from './apiEndpoint';
export * from './appointment';
export * from './common';
export * from './firebase';
export * from './image';
export * from './message';
export * from './navigation';
export * from './pagination';
export * from './regex';
export * from './router';
export * from './timing';
export * from './chemist';
export * from './encode';
