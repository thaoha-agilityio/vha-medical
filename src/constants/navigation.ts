import {
  AppointmentIcon,
  ChemistIcon,
  DashboardIcon,
  DoctorIcon,
  HospitalIcon,
  MedicineIcon,
  MessageIcon,
  SettingIcon,
} from '@/icons';

// Router
import { PRIVATE_ROUTES } from './router';

export const NAVBAR_LINKS = [
  {
    name: 'Dashboard',
    href: PRIVATE_ROUTES.DASHBOARD,
    icon: DashboardIcon,
    isEnable: true,
  },
  {
    name: 'Appointments',
    href: PRIVATE_ROUTES.APPOINTMENTS,
    icon: AppointmentIcon,
    isEnable: true,
  },
  {
    name: 'Chemists',
    href: PRIVATE_ROUTES.CHEMISTS,
    icon: ChemistIcon,
    isEnable: true,
  },
  { name: 'Hospitals', href: '#', icon: HospitalIcon },
  { name: 'Doctors', href: '#', icon: DoctorIcon },
  { name: 'Medicines', href: '#', icon: MedicineIcon },
];

export const SUPPORT_LINKS = [
  { name: 'Messages', href: '#', icon: MessageIcon },
  {
    name: 'Settings',
    href: '#',
    icon: SettingIcon,
  },
];
