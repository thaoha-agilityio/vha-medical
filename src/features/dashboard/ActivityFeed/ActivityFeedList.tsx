'use client';

import { lazy, memo, useTransition } from 'react';
import { Card } from '@nextui-org/react';

// Types
import {
  APIRelatedResponse,
  APIResponse,
  ColumnType,
  InfoModel,
  MetaResponse,
  NotificationModel,
  NotificationResponse,
  UserModel,
} from '@/types';

// Components
import { Avatar, Status, Text } from '@/components/ui';
import {
  formatDateTime,
  fromDateToNow,
  getDescriptionNotification,
} from '@/utils';
import { ActivityFeedListSkeleton } from './ActivityFeedSkeleton';

const DataGrid = lazy(() => import('@/components/ui/DataGrid'));

interface ActivityInfoProps {
  item: NotificationModel;
  userId?: string;
}

const ActivityInfo = memo(({ item, userId = '' }: ActivityInfoProps) => {
  const {
    senderAvatar = '',
    createdAt = '',
    info = {} as InfoModel,
    senderId = {} as APIRelatedResponse<APIResponse<UserModel>>,
  } = item || {};
  const { startTime = '', content } = info;

  const time = formatDateTime(startTime);
  const description = getDescriptionNotification({
    userId,
    senderId,
    time,
    content,
  });

  const timeAgo = fromDateToNow(createdAt);

  return (
    <div className="flex gap-3 justify-items-start ml-1">
      <Avatar isBordered src={senderAvatar} className="aspect-square" />
      <div className="flex flex-col mr-8">
        <Text variant="description" customClass="text-2xs md:text-xs text-wrap">
          {description}
        </Text>
        <Text variant="subTitle" size="2xs">
          {timeAgo}
        </Text>
      </div>
    </div>
  );
});

export const COLUMNS_ACTIVITY_FEED: ColumnType<NotificationModel>[] = [
  {
    key: 'sender',
    title: 'Sender',
  },
  {
    key: 'status',
    title: 'Status',
    customNode: ({ item }) => {
      const { info } = item || {};
      const { status = 0 } = info || {};

      return <Status status={status} className="leading-[27px] max-w-[70px]" />;
    },
  },
];
export interface ActivityFeedListProps extends MetaResponse {
  notifications: NotificationResponse[];
  userId: string;
}

const ActivityFeedList = memo(
  ({ userId, notifications, pagination }: ActivityFeedListProps) => {
    const [isPending, startTransition] = useTransition();

    return (
      <Card as="section" className="bg-background-200 h-fit w-full p-4 md:py-5">
        <Text variant="title" size="lg" customClass="leading-9">
          Activity Feed
        </Text>
        {isPending ? (
          <ActivityFeedListSkeleton />
        ) : (
          <div className="flex flex-col items-center">
            <DataGrid
              id="activity-feed"
              data={notifications}
              pagination={pagination}
              startTransition={startTransition}
              columns={
                COLUMNS_ACTIVITY_FEED.map((column) => ({
                  ...column,
                  customNode: ({ item }) =>
                    column.key === 'sender' ? (
                      <ActivityInfo
                        item={item as NotificationModel}
                        userId={userId}
                      />
                    ) : column.customNode ? (
                      column.customNode({
                        column,
                        item: item as NotificationModel,
                      })
                    ) : null,
                })) as ColumnType<unknown>[]
              }
              classWrapper="pt-4"
              classCell="pb-6"
            />
          </div>
        )}
      </Card>
    );
  },
);

ActivityFeedList.displayName = 'ActivityFeedList';
ActivityInfo.displayName = 'ActivityInfo';
export default ActivityFeedList;
