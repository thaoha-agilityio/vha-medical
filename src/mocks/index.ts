export * from './appointments';
export * from './auth';
export * from './chemists';
export * from './common';
export * from './dateTime';
export * from './notification';
export * from './options';
export * from './profile';
export * from './table';
export * from './user';
