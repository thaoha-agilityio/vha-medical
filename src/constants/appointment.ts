import { AppointmentModel, AppointmentStatusOption } from '@/types';
import { DEFAULT_CHEMIST_DATA } from './chemist';

export const APPOINTMENT_STATUS_OPTIONS: AppointmentStatusOption[] = [
  {
    key: 'new',
    label: 'New',
    value: 0,
  },
  {
    key: 'meeting',
    label: 'Meeting',
    value: 1,
  },
  {
    key: 'cancelled',
    label: 'Cancelled',
    value: 2,
  },
];

export const APPOINTMENT_STATUS = [
  {
    key: '0',
    label: 'New',
  },
  {
    key: '1',
    label: 'Meeting',
  },
  {
    key: '2',
    label: 'Canceled',
  },
];

export const DEFAULT_APPOINTMENT_DATA: AppointmentModel = {
  startTime: '',
  durationTime: '',
  status: 0,
  receiverId: {
    data: {
      id: '',
      attributes: DEFAULT_CHEMIST_DATA,
    },
  },
  senderId: {
    data: {
      id: '',
      attributes: DEFAULT_CHEMIST_DATA,
    },
  },
};

export const APPOINTMENT_SEARCH_PARAMS = ['receiverId', 'senderId'] as const;
