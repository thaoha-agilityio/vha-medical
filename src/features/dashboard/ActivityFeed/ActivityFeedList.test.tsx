import { render } from '@testing-library/react';
import React, { act } from 'react';
import { usePathname, useRouter, useSearchParams } from 'next/navigation';

// Mocks
import { MOCK_NOTIFICATION_LIST } from '@/mocks';

// Components
import ActivityFeedList, { ActivityFeedListProps } from './ActivityFeedList';

// Constants
import { PRIVATE_ROUTES } from '@/constants';

jest.mock('next/navigation', () => ({
  usePathname: jest.fn(),
  useSearchParams: jest.fn(),
  useRouter: jest.fn(),
}));

describe('ActivityFeedList test cases', () => {
  const mockProps: ActivityFeedListProps = {
    notifications: MOCK_NOTIFICATION_LIST,
    userId: '1',
    pagination: {
      page: 1,
      pageSize: 10,
      pageCount: 1,
      total: 3,
    },
  };
  const mockReplace = jest.fn();
  const mockUsePathname = usePathname as jest.Mock;
  const mockUseRouter = useRouter as jest.Mock;
  const mockUseSearchParams = useSearchParams as jest.Mock;

  beforeEach(() => {
    mockUsePathname.mockReturnValue(PRIVATE_ROUTES.DASHBOARD);
    mockUseRouter.mockReturnValue({ replace: mockReplace });
    mockUseSearchParams.mockReturnValue(new URLSearchParams());
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  const setup = async (props: ActivityFeedListProps) =>
    await act(() => render(<ActivityFeedList {...props} />));

  it('should render correctly', async () => {
    const { asFragment } = await setup(mockProps);

    expect(asFragment()).toMatchSnapshot();
  });
});
