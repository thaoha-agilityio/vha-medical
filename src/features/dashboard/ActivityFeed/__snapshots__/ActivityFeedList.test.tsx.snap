// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityFeedList test cases should render correctly 1`] = `
<DocumentFragment>
  <section
    class="flex flex-col relative overflow-hidden text-foreground box-border outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-large transition-transform-background motion-reduce:transition-none bg-background-200 h-fit w-full p-4 md:py-5"
    tabindex="-1"
  >
    <p
      class="text-primary-100 font-semibold text-lg leading-9"
    >
      Activity Feed
    </p>
    <div
      class="flex flex-col items-center"
    >
      <div
        class="flex flex-col relative gap-4 w-full"
      >
        <div
          class="z-0 flex flex-col relative justify-between gap-4 overflow-auto rounded-large w-full bg-transparent-200 shadow-none p-0 pt-4"
        >
          <table
            aria-describedby=""
            aria-label="Table"
            class="min-w-full h-auto table-auto w-full"
            id="activity-feed"
            role="grid"
            tabindex="0"
          >
            <thead
              class="[&>tr]:first:rounded-lg hidden"
              role="rowgroup"
            >
              <tr
                class="group outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2"
                role="row"
              >
                <th
                  class="group px-3 h-10 align-middle bg-default-100 whitespace-nowrap text-foreground-500 text-tiny font-semibold first:rounded-s-lg last:rounded-e-lg data-[sortable=true]:cursor-pointer data-[hover=true]:text-foreground-400 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-start"
                  data-key="sender"
                  id="activity-feed-sender"
                  role="columnheader"
                  tabindex="-1"
                >
                  Sender
                </th>
                <th
                  class="group px-3 h-10 align-middle bg-default-100 whitespace-nowrap text-foreground-500 text-tiny font-semibold first:rounded-s-lg last:rounded-e-lg data-[sortable=true]:cursor-pointer data-[hover=true]:text-foreground-400 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-start"
                  data-key="status"
                  id="activity-feed-status"
                  role="columnheader"
                  tabindex="-1"
                >
                  Status
                </th>
              </tr>
              <tr
                aria-hidden="true"
                class="w-px h-px block"
                style="margin-left: 0.25rem; margin-top: 0.25rem;"
                tabindex="-1"
              />
            </thead>
            <tbody
              role="rowgroup"
            >
              <tr
                aria-labelledby="activity-feed-1-sender"
                class="group outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2"
                data-first="true"
                data-key="1"
                role="row"
                tabindex="-1"
              >
                <td
                  class="relative align-middle whitespace-normal text-small font-normal [&>*]:z-1 [&>*]:relative outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 before:content-[''] before:absolute before:z-0 before:inset-0 before:opacity-0 data-[selected=true]:before:opacity-100 group-data-[disabled=true]:text-foreground-300 group-data-[disabled=true]:cursor-not-allowed before:bg-default/60 data-[selected=true]:text-default-foreground first:before:rounded-s-lg last:before:rounded-e-lg text-start p-0 pb-6"
                  data-key="table-row-cell-sender"
                  headers="activity-feed-sender"
                  id="activity-feed-1-sender"
                  role="cell"
                  tabindex="-1"
                >
                  <div
                    class="flex gap-3 justify-items-start ml-1"
                  >
                    <span
                      class="overflow-hidden outline-none rounded-full ring-offset-background dark:ring-offset-background-dark shrink-0 min-w-8 w-8 h-8 bg-undefined ring-offset-2 ring-1 ring-green false"
                      data-testid="avatar-container"
                      tabindex="-1"
                    >
                      <img
                        alt=""
                        class="aspect-square"
                        data-nimg="1"
                        decoding="async"
                        fetchpriority="high"
                        height="48"
                        src="http://localhost/_next/image?url=%2Fupload%2F300&w=96&q=75"
                        srcset="/_next/image?url=%2Fupload%2F300&w=48&q=75 1x, /_next/image?url=%2Fupload%2F300&w=96&q=75 2x"
                        style="color: transparent; object-fit: cover; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
                        width="48"
                      />
                    </span>
                    <div
                      class="flex flex-col mr-8"
                    >
                      <p
                        class="text-primary-200 font-medium text-2xs md:text-xs text-wrap"
                      >
                         updated at Sep 12, 2024 9:00 am
                      </p>
                      <p
                        class="text-primary-300 font-normal text-2xs"
                      >
                        a month ago
                      </p>
                    </div>
                  </div>
                </td>
                <td
                  class="relative align-middle whitespace-normal text-small font-normal [&>*]:z-1 [&>*]:relative outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 before:content-[''] before:absolute before:z-0 before:inset-0 before:opacity-0 data-[selected=true]:before:opacity-100 group-data-[disabled=true]:text-foreground-300 group-data-[disabled=true]:cursor-not-allowed before:bg-default/60 data-[selected=true]:text-default-foreground first:before:rounded-s-lg last:before:rounded-e-lg text-start p-0 pb-6"
                  data-key="table-row-cell-status"
                  headers="activity-feed-status"
                  role="cell"
                  tabindex="-1"
                >
                  <div
                    class="bg-success text-white dark:text-success dark:bg-opacity-20 w-[68px] px-2 rounded-md text-center leading-[27px] max-w-[70px]"
                  >
                    <p
                      class="text-2xs font-semibold capitalize"
                    >
                      new
                    </p>
                  </div>
                </td>
              </tr>
              <tr
                aria-labelledby="activity-feed-2-sender"
                class="group outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2"
                data-key="2"
                data-last="true"
                data-odd="true"
                role="row"
                tabindex="-1"
              >
                <td
                  class="relative align-middle whitespace-normal text-small font-normal [&>*]:z-1 [&>*]:relative outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 before:content-[''] before:absolute before:z-0 before:inset-0 before:opacity-0 data-[selected=true]:before:opacity-100 group-data-[disabled=true]:text-foreground-300 group-data-[disabled=true]:cursor-not-allowed before:bg-default/60 data-[selected=true]:text-default-foreground first:before:rounded-s-lg last:before:rounded-e-lg text-start p-0"
                  data-key="table-row-cell-sender"
                  headers="activity-feed-sender"
                  id="activity-feed-2-sender"
                  role="cell"
                  tabindex="-1"
                >
                  <div
                    class="flex gap-3 justify-items-start ml-1"
                  >
                    <span
                      class="overflow-hidden outline-none rounded-full ring-offset-background dark:ring-offset-background-dark shrink-0 min-w-8 w-8 h-8 bg-undefined ring-offset-2 ring-1 ring-green false"
                      data-testid="avatar-container"
                      tabindex="-1"
                    >
                      <img
                        alt=""
                        class="aspect-square"
                        data-nimg="1"
                        decoding="async"
                        fetchpriority="high"
                        height="48"
                        src="http://localhost/_next/image?url=%2Fupload%2F300&w=96&q=75"
                        srcset="/_next/image?url=%2Fupload%2F300&w=48&q=75 1x, /_next/image?url=%2Fupload%2F300&w=96&q=75 2x"
                        style="color: transparent; object-fit: cover; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
                        width="48"
                      />
                    </span>
                    <div
                      class="flex flex-col mr-8"
                    >
                      <p
                        class="text-primary-200 font-medium text-2xs md:text-xs text-wrap"
                      >
                         deleted at Sep 12, 2024 9:00 am
                      </p>
                      <p
                        class="text-primary-300 font-normal text-2xs"
                      >
                        a month ago
                      </p>
                    </div>
                  </div>
                </td>
                <td
                  class="relative align-middle whitespace-normal text-small font-normal [&>*]:z-1 [&>*]:relative outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 before:content-[''] before:absolute before:z-0 before:inset-0 before:opacity-0 data-[selected=true]:before:opacity-100 group-data-[disabled=true]:text-foreground-300 group-data-[disabled=true]:cursor-not-allowed before:bg-default/60 data-[selected=true]:text-default-foreground first:before:rounded-s-lg last:before:rounded-e-lg text-start p-0"
                  data-key="table-row-cell-status"
                  headers="activity-feed-status"
                  role="cell"
                  tabindex="-1"
                >
                  <div
                    class="bg-success text-white dark:text-success dark:bg-opacity-20 w-[68px] px-2 rounded-md text-center leading-[27px] max-w-[70px]"
                  >
                    <p
                      class="text-2xs font-semibold capitalize"
                    >
                      new
                    </p>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
